docs/coding-conventions.md

```diff
@@ -0,0 +1,297 @@
+    obj.GetAnnotations()[annotationKey] = val
+}
+```
+
+# Uniform import order and naming
+
+- Use the following import order with one block per item:
+    - Golang standard libraries
+    - Ginkgo / Gomega imports (only in test files)
+    - Third-party libraries
+    - Local packages
+- Use the following naming schemes for imports:
+    - `virtv1` for imports of `kubevirt.io/api`
+    - `metav1` for imports of `k8s.io/apimachinery/pkg/apis/meta/v1`
+    - `k8s` prefix for `k8s.io` imports on collision with other imports
+    - TODO: Add more
+- `make format` is able to help you with this.
+
+### Bad example
+
+```go
+import (
+    "context"
+    k8sv1 "k8s.io/api/core/v1"
+    "time"
+
+    ."github.com/onsi/ginkgo/v2"
+    ."github.com/onsi/gomega"
+    metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+    v1 "kubevirt.io/api/core/v1"
+    "kubevirt.io/kubevirt/tests/clientcmd"
+
+    "kubevirt.io/kubevirt/tests/testsuite"
+)
+```
+
+### Good example
+
+```go
+import (
+    "context"
+    "time"
+
+    ."github.com/onsi/ginkgo/v2"
+    ."github.com/onsi/gomega"
+
+    k8sv1 "k8s.io/api/core/v1"
+    metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+    virtv1 "kubevirt.io/api/core/v1"
+
+    "kubevirt.io/kubevirt/tests/clientcmd"
+    "kubevirt.io/kubevirt/tests/testsuite"
+)
+```
+
+## Short coding conventions (TLDR)
+
+Here are some important but short conventions. If you want to learn more
+about conventions in detail take a look at the [detailed
+coding conventions](#detailed-coding-conventions).
+
+- [Write elegant, cohesive and easily readable code](#write-elegant-cohesive-and-easily-readable-code)
+- [Avoid nesting and complexity by using early returns](#avoid-nesting-and-complexity-by-using-early-returns)
+- [Use the same coding style throughout a file](#use-the-same-coding-style-throughout-a-file)
+- [Avoid repetition of hardcoded values](#avoid-repetition-of-hardcoded-values)
+- [Uniform import order and naming](#uniform-import-order-and-naming)
+- Prefert to inline variable assignments.
+    - For example: Inline `err` checks in if-statements.
+- Use switch-cases to avoid long if/else if/else statements.
+- Isolate code.
+    - Use interfaces instead of bare structs.
+- Avoid use of global variables.
+    - Use structs and receiver methods to keep state.
+- Avoid long files.
+    - Avoid adding helpers in a single place. Long files like `tests/utils.go`
+      are hard to maintain.
+    - Alternatively, add helpers to the places where they are used or group them
+      in packages/files with representative names.
+- Prefer to define variables in the function body
+    - Avoid returning too many values from a function.
+    - Avoid naked returns.
+- Use closures with caution, be aware of the risks and use them only when it
+  makes sense.
+    - They can still be used if there is a good reason for it.
+    - A useful use case of closures is to define a function that is relevant
+      only for a very specific scope.
```

Comment:
I do not understand this one.

How about this:
> - Defined and used inline, not though variable assignment.
> - Used as actual closures, binding to external variables.
> - Grouping a set of instructions to provide common scope functionality (e.g. `defer`).

Reply:
Discussion about this started here: https://github.com/kubevirt/kubevirt/pull/11456#discussion_r1515623523

Do you want me to add these items as examples?

Reply:
The current text is generic, leaving too much opening to do all kind of odd things.

I am open to negotiate with others about specific cases.
Is there a case that makes sense to use them that is not in the list I mentioned?

If the point is to leave something open for exceptions, lets say so explicitly.
E.g.:
> - Exceptions may exist, but they need to be well reasoned for, expressing what alternatives have been considered.

Reply:
ACK, I see, updated it.